lib/backend/consul/consul.go

```diff
@@ -0,0 +1,613 @@
+	ConsulOp     *consulapi.KVTxnOp
+	DefaultError error
+}
+
+type setOperationKind int
+
+const (
+	create setOperationKind = iota
+	replace
+	update
+)
+
+type setOptions struct {
+	Kind  setOperationKind
+	Index uint64
+}
+
+func NewConsulClient(config *ConsulConfig) (*ClientWrapper, error) {
+	auth := consulapi.HttpBasicAuth{
+		Password: config.ConsulPassword,
+		Username: config.ConsulUsername,
+	}
+	if len(config.ConsulAddress) == 0 {
+		return nil, goerrors.New("no consul address provided")
+	}
+
+	cfg := consulapi.Config{
+		Address:    config.ConsulAddress,
+		Scheme:     config.ConsulScheme,
+		Datacenter: config.ConsulDatacenter,
+		Token:      config.ConsulToken,
+		HttpAuth:   &auth,
+		WaitTime:   watchTimeout,
+	}
+
+	client, err := consulapi.NewClient(&cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &ClientWrapper{Client: client}, nil
+}
+
+func getReadyFlagPair() *model.KVPair {
+	return &model.KVPair{
+		Key:   model.ReadyFlagKey{},
+		Value: true,
+	}
+}
+
+// EnsureInitialized makes sure that the etcd data is initialized for use by
+// Calico.
+func (c *ClientWrapper) EnsureInitialized() error {
+	// Make sure the Ready flag is initialized in the datastore
+	if _, err := c.Create(getReadyFlagPair()); err == nil {
+		log.Info("Ready flag is now set")
+	} else {
+		if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+			log.WithError(err).Warn("Failed to set ready flag")
+			return err
+		}
+		log.Info("Ready flag is already set")
+	}
+
+	return nil
+}
+
+func (c *ClientWrapper) Syncer(callbacks api.SyncerCallbacks) api.Syncer {
+	return newSyncer(c.Client, callbacks)
+}
+
+// Create an entry in the datastore.  This errors if the entry already exists.
+func (c *ClientWrapper) Create(d *model.KVPair) (*model.KVPair, error) {
+	return c.set(d, &setOptions{Kind: create})
+}
+
+// Update an existing entry in the datastore.  This errors if the entry does
+// not exist.
+func (c *ClientWrapper) Update(d *model.KVPair) (*model.KVPair, error) {
+	// If the request includes a revision, set it as the etcd previous index.
+	options := setOptions{Kind: update}
+	if d.Revision != nil {
+		options.Index = d.Revision.(uint64)
+		log.Debugf("Performing CAS against consulapi index: %v\n", options.Index)
+	}
+
+	return c.set(d, &options)
+}
+
+// Set an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *ClientWrapper) Apply(d *model.KVPair) (*model.KVPair, error) {
+	return c.set(d, &setOptions{Kind: replace})
+}
+
+// Delete an entry in the datastore.  This errors if the entry does not exists.
+func (c *ClientWrapper) Delete(d *model.KVPair) error {
+	path, err := keyToDefaultDeletePath(d.Key)
+	if err != nil {
+		return err
+	}
+	deleteOp := &consulapi.KVTxnOp{Key: path, Verb: consulapi.KVDeleteTree}
+	if d.Revision != nil {
+		deleteOp.Index = d.Revision.(uint64)
+	}
+	log.Debugf("Delete Key: %s", path)
+
+	kv := c.Client.KV()
+
+	readyFlagPair := getReadyFlagPair()
```

Comment:
(I've looked through the rest of the code / the PR itself, but it should be documented in the comments.)